packages/panzer/dof-mgr/test/fe_assembly/test_fe_assembly_HEX.hpp

```diff
@@ -0,0 +1,608 @@
+      dofManager->getElementGIDs(elem_id, elementGIDs);
+
+
+      for(ordinal_type node_id=0; node_id<basisCardinality; node_id++)
+        column_local_ids[node_id] = fe_matrix->getColMap()->getLocalElement(elementGIDs[elmtOffset[node_id]]);
+
+      // For each node (row) on the current element:
+      // - populate the values array
+      // - add the values to the fe_matrix.
+
+      for(ordinal_type node_id=0; node_id<basisCardinality; node_id++)
+      {
+        global_ordinal_t local_row_id = elementLIDsHost(elem_id, elmtOffset[node_id]);
+
+        for(ordinal_type col_idx=0; col_idx<basisCardinality; col_idx++)
+          column_scalar_values[col_idx] = cellMatHost(elem_id, node_id, col_idx);
+
+        fe_matrix->sumIntoLocalValues(local_row_id, column_local_ids, column_scalar_values);
+        rhs->sumIntoLocalValue(column_local_ids[node_id], 0, cellRhsHost(elem_id, node_id));
+      }
+    }
+
+    Tpetra::endFill(*fe_matrix);
+    Tpetra::endFill(*rhs);
+
+
+
+
+    // ************************************ CODE VERIFICATION **************************************
+
+
+    DynRankView ConstructWithLabel(basisCoeffsLI, numOwnedElems, basisCardinality);
+    {
+      DynRankView ConstructWithLabel(dofCoordsOriented, numOwnedElems, basisCardinality, dim);
+      DynRankView ConstructWithLabel(dofCoeffsPhys, numOwnedElems, basisCardinality);
+
+      li::getDofCoordsAndCoeffs(dofCoordsOriented,  dofCoeffsPhys, basis.getRawPtr(), POINTTYPE_EQUISPACED, elemOrts);
+
+      //Compute physical Dof Coordinates
+
+      DynRankView ConstructWithLabel(funAtDofPoints, numOwnedElems, basisCardinality);
+      {
+        DynRankView ConstructWithLabel(physDofPoints, numOwnedElems, numQPoints, dim);
+        ct::mapToPhysicalFrame(physDofPoints,dofCoordsOriented,physVertexes,basis->getBaseCellTopology());
+        EvalSolFunctor<DynRankView> functor;
+        functor.funAtPoints = funAtDofPoints;
+        functor.points = physDofPoints;
+        Kokkos::parallel_for(numOwnedElems,functor);
+      }
+
+      li::getBasisCoeffs(basisCoeffsLI, funAtDofPoints, dofCoeffsPhys);
+    }
+
+    {
+      Tpetra::Vector<scalar_t> sol(domain_map);
+      auto basisCoeffsLIHost = Kokkos::create_mirror_view(basisCoeffsLI);
+      Kokkos::deep_copy(basisCoeffsLIHost,basisCoeffsLI);
+      for(ordinal_type elem_id=0; elem_id<numOwnedElems; elem_id++)
+      {
+        //auto elementLIDs = ugi->getElementLIDs(elem_id);
+        dofManager->getElementGIDs(elem_id, elementGIDs);
+
+
+        for(ordinal_type node_id=0; node_id<basisCardinality; node_id++)
+        {
+          global_ordinal_t gid = elementGIDs[elmtOffset[node_id]];
+          if(domain_map->isNodeGlobalElement(gid))
+            sol.replaceGlobalValue(gid, basisCoeffsLIHost(elem_id, node_id));
+        }
+      }
+
+      fe_matrix->apply(sol, *rhs, Teuchos::NO_TRANS, -1.0, 1.0);   // b - A x
+    }
+
+    *outStream << "Residual norm : " << rhs->getVector(0)->norm2() <<std::endl;
```

Comment:
Another quick question - how would this test fail in ctest for an incorrectly assembled matrix/rhs? It looks like the check prints the residual norm to screen instead of checking a tolerance and throwing.

Reply:
@rppawlo yeah, I plan to check if the error is within some tolerance. In principle one can change the order of the basis and number of elements as arguments of the executable. I can add a check only when the parameters are the default ones.

Reply:
Thanks @rppawlo I'll add the fences.